src/Map.php

```diff
@@ -25,11 +25,25 @@ final class Map implements Collection, \ArrayAccess
     const MIN_CAPACITY = 8;
 
     /**
-     * @var array internal array to store pairs
+     * @var int internal ordered index for the next pair ref.
+     */
+    private $nextPairIndex = 0;
+
+    /**
+     * @var array internal array to store ordered references to pairs. Although
+     * the array has ascending integer indices, they aren't necessarily
+     * contiguous.
      *
-     * @psalm-var array<int, Pair>
+     * @psalm-var array<int, PairRef>
      */
-    private $pairs = [];
+    private $pairRefs = [];
+
+    /**
+     * @var array internal lookup table to quickly find pairs by hash key.
+     *
+     * @psalm-var array<array-key, list<PairRef>>
+     */
+    private $table = [];
```

Comment:
i feel like this coudl potentiallly be reduced to PairRef->key and then use that to lookup in the pairRefs table

Reply:
I'm not quite following. The `pairRefs` array doesn't have (stable, meaningful) keys. It maintains the order of elements in the map.

Reply:
ah i see i thought the keys were the hashes for some reason